### ISSUE TITLE
[우아한 타입스크립트 with 리액트] 3.1 타입스크립트만의 독자적 타입 시스템 (1)

### ISSUE THREAD

**Issue by BangDori** [BangDori/woowa-typescript#19]
# Enum 타입

enum 타입은 각 멤버에 대한 값을 명시하지 않아도 자동적으로 값이 할당됩니다.

```
enum ProgrammingLanguage {
  typescript, // 0
  javascript, // 1
  java, // 2
  python, // 3
}
```

위에서 `typescript`의 값에 대해 접근하기 위해 `ProgrammingLanguage.typescript`를 사용할 수도 있지만, **역방향으로도 접근이 가능**하다는 특징이 있습니다.

뿐만 아니라 선생님이 #16 에서 작성해주신 **UserRole** 의 예시와 같이, `enum`을 잘 활용한다면 코드의 가독성을 크게 향상시켜줄 수 있을 것 같습니다.

```
enum ItemStatusType {
  DELIVERY_HOLD = "DELIVERY_HOLD", // 배송 보류
  DELIVERY_READY = "DELIVERY_READY", // 배송 준비 중
  DELIVERING = "DELIVERING", // 배송 중
  DELIVERED = "DELIVERED", // 배송 완료
}

const checkItemAvailable = (itemStatus: ItemStatusType) => {
  switch (itemStatus) {
    case ItemStatusType.DELIVERY_HOLD:
    case ItemStatusType.DELIVERY_READY:
    case ItemStatusType.DELIVERING:
    case ItemStatusType.DELIVERED:
  }
};
```

그리고 `enum`에 대한 설명을 덧붙이자면 `itemStatusType` 열거형을 타입으로 가짐으로써 얻을 수 있는 효과는 아래와 같습니다.

1. **타입 안정성**: ItemStatusType에 명시되지 않은 다른 문자열은 인자로 받을 수 없다. 따라서 타입 안정성이 우수하다.
2. **명확한 의미 전달과 높은 응집력**: ItemStatusType 타입이 다루는 값이 무엇인지 명확하다. **아이템 상태에 대한 값을 모아놓은 것으로 응집력이 뛰어나다**.
3. **가독성**: 응집도가 높기 때문에 말하고자 하는 바가 더욱 명확하다. 따라서 열거형 멤버를 통해 어떤 상태를 나타내는지 쉽게 이해할 수 있다. (예: `ItemStatusType.DELIVERY_HOLD vs DELIVERY_HOL`)

이처럼 **열거형은 관련이 높은 멤버를 모아 문자열 상수처럼 사용하고자 할 때 유용**하게 쓸 수 있다.

다만 열거형에 사용할 때는 주의해야 할 점이 있다. 먼저 **숫자로만 이루어져 있거나 타입스크립트가 자동으로 추론한 열거형은 안전하지 않은 결과**를 낳을 수 있다. 맨 처음 예시를 보면 역방향으로도 접근할 수 있음을 보여준다. 여기서 할당된 값을 넘어서는 범위로 역방향으로 접근하더라도 타입스크립트는 막지 않는다.

이러한 동작을 막기 위해 **`const enum`으로 열거형을 선언**하는 방법이 있다. 이 방식은 역방향으로의 접근을 허용하지 않기 때문에 자바스크립트에서의 객체에 접근하는 것과 유사한 동작을 보장한다.

```
ProgrammingLanguage[200]; // undefined를 출력하지만 별다른 에러를 발생시키지 않는다.

// 다음과 같이 선언하면 위와 같은 문제를 방지할 수 있다
const enum ProgrammingLanguage {
   ...
}
```

그러나 `const enum`으로 열거형을 선언하더라도 숫자 상수로 관리되는 열거형은 선언한 값 이외의 값을 할당하거나 접근할 때 이를 방지하지 못한다. 반면 문자열 상수 방식으로 선언한 열거형은 미리 선언하지 않은 멤버로 접근을 방지한다. 따라서 문자열 상수 방식으로 열거형을 사용하는 것이 숫자 상수 방식보다 더 안전하며 의도하지 않은 값의 할당이나 접근을 방지하는 데 도움이 된다.

```
const enum NUMBER {
  ONE = 1,
  TWO = 2,
}

const myNumber: NUMBER = 100; // NUMBER enum에서 100을 관리하고 있지 않지만 이는 에러를 발생시키지 않는다.

const enum STRING_NUMBER {
  ONE = "ONE",
  TWO = "TWO",
}

const myStringNumber: STRING_NUMBER = "THREE"; // ERROR
```

이외에도 열거형의 가장 큰 문제는 따로 존재한다. 앞서 열거형은 타입 공간과 값 공간에서 모두 사용된다고 말했다. 해당 예시에서 열거형은 타입스크립트 코드가 자바스크립트로 변환될 때 **즉시 실행 함수(IIFE)형식으로 변환**되는 것을 볼 수 있다.

<img src="https://github.com/BangDori/woowa-typescript/assets/44726494/be39ce68-603c-428b-93b8-25c3d1e5ca61" width="360" />
<img src="https://github.com/BangDori/woowa-typescript/assets/44726494/4ca5d3ac-a65c-484d-91fd-cccce0117ee3" width="330" />

이때 일부 번들러에서 **트리쉐이킹** 과정 중 즉시 실행 함수로 변환된 값을 사용하지 않는 코드로 인식하지 못하는 경우가 발생할 수 있다. 따라서 언급했던 `const enum` 또는 `as const assertion`을 사용해서 유니온 타입으로 열거형과 동일한 효과를 얻는 방법이 있다.

**트리쉐이킹, 즉시 실행 함수(IIFE)와 즉시 실행 함수로 변환되는 것이 왜 단점이 될 지**에 대해서는 [우아한 타입스크립트 with 리액트] 3.1 타입스크립트만의 독자적 타입 시스템 (2) 에서 작성을 이어나가겠습니다.

---

@youngsu5582 선생님 위 코드를 타이핑하였을 때, `const myNumber: NUMBER = 100` 부분에서 NUMBER 내부에 100이라는 값을 가지고 있지 않아 에러가 발생하는 것을 확인하였는데 이는 책 내용에서 오류가 있는 것이겠죠?

**Comment by youngsu5582** [comment]
해당 부분은 아마도 , 책에서 오류가 있는거 같습니다!
const enum 과 enum의 이런 차이점이 있는지 몰랐네요!
하지만 , 두가지를 잘 고려해서 사용해야 할 거 같습니다
말하신 것 처럼 const enum 으로 선언할 시 , 컴파일 때 사라지게 됩니다.

코드 작성중에만 도움을 받을때는 상관이 없으나,
```javascript
enum STRING_NUMBER {
  ONE = "ONE1",
  TWO = "TWO2",
}

// Iterating over the keys of the enum
for (let key of Object.keys(STRING_NUMBER)) {
  console.log(key);
}
```
해당 코드에서 , const enum 으로 변환 시 enum이 사라지게 되므로
> 'const' enums can only be used in property or index access expressions or the right hand side of an import declaration or export assignment or type query.
해당 에러가 발생하게 됩니다.

물론 , 매우 특이한 경우이며 이런 경우도 있다 생각만 하시면 될 거 같습니다😊!

**Comment by BangDori** [comment]
맞습니다. `const enum`으로 선언한 후에 자바스크립트로 컴파일하면 `const enum`으로 선언한 부분은 사라지게 되더라구요.

![image](https://github.com/BangDori/woowa-typescript/assets/44726494/0b6084b6-e051-40b0-ad7f-23ba38eb5587)

```
const enum NUMBER {
  ONE = 1,
  TWO = 2,
  THREE = 3,
}

const number: NUMBER = 1;
// 🚨'const' enums can only be used in property or index access expressions
// or the right hand side of an import declaration or export assignment or type query.
for (let key of Object.keys(NUMBER)) {
  console.log(key);
}

console.log(number);
```

선생님이 말씀해주신 것처럼 `enum` 자체를 순회하는 경우 에러가 발생하지만, 이는 매우 특이한 케이스일 것 같네요!

`const enum`을 무작정 채택하면 좋을 것 같다는 생각을 했는데 짚어주셔서 감사합니다!
